src/notifications/api/serializers.py

```diff
@@ -1,19 +1,197 @@
 import logging
+from json import dumps
 
+from django.core.exceptions import ObjectDoesNotExist
+from django.core.serializers.json import DjangoJSONEncoder
+
+import requests
 from rest_framework import serializers
+from rest_framework.response import Response
+
+from notifications.conf.base import CHANNEL
+from notifications.datamodel.models import (
+    Abonnement, Filter, FilterGroup, Kanaal
+)
 
 logger = logging.getLogger(__name__)
 
 
-class ExampleSerializer(serializers.HyperlinkedModelSerializer):
+class FilterSerializer(serializers.ModelSerializer):
+
+    def to_representation(self, instance):
+        data = super().to_representation(instance)
+        return {data['key']: data['value']}
+
+    def to_internal_value(self, data):
+        if len(data) > 1:
+            raise serializers.ValidationError(
+                {'filter': "'filter dict must have only one element')"},
+                code='filter_many'
+            )
+        key = list(data)[0]
+        value = data[key]
+        return Filter(key=key, value=value)
+
+    class Meta:
+        model = Filter
+        fields = (
+            'key',
+            'value',
+        )
+
+
+class KanaalSerializer(serializers.ModelSerializer):
+
     class Meta:
-        model = ...
+        model = Kanaal
         fields = (
             'url',
-            ...
+            'naam',
+            'documentatie_link',
         )
-        extra_kwarsg = {
+        extra_kwargs = {
             'url': {
                 'lookup_field': 'uuid',
             },
+            'documentatie_link': {
+                'required': False,
+            }
+        }
+
+
+class FilterGroupSerializer(serializers.ModelSerializer):
+    filters = FilterSerializer(many=True, required=False)
+    naam = serializers.CharField(source='kanaal.naam')
+
+    class Meta:
+        model = FilterGroup
+        fields = (
+            'filters',
+            'naam',
+        )
+
+        # delete unique validator for naam field - process it manually in AbonnementSerializer.create()
+        extra_kwargs = {
+            'naam': {
+                'validators': [],
+            }
         }
+
+
+class AbonnementSerializer(serializers.HyperlinkedModelSerializer):
+    kanalen = FilterGroupSerializer(source='filter_groups', many=True)
+
+    class Meta:
+        model = Abonnement
+        fields = (
+            'url',
+            'callback_url',
+            'auth',
+            'kanalen',
+        )
+        extra_kwargs = {
+            'url': {
+                'lookup_field': 'uuid',
+            },
+            'auth': {
+                'write_only': True,
+            }
+        }
+
+    def validate(self, attrs):
+        validated_attrs = super().validate(attrs)
+        for group_data in validated_attrs['filter_groups']:
```

Comment:
Why is this in the `AbonnementSerializer` and not in the `FilterGroupSerializer`?

No immediate need to change it, but honest question.

Reply:
> Anna mentioned due to a bug in vng-api-common, nested serialization goes wrong if it fails.

We choose to be pragmatic and keep it as is, but would be nice if vng-api-common was updated :D